0000-modules/detailed-design/loading-files.md

```diff
@@ -0,0 +1,194 @@
+
+Test, example, and benchmark crates follow a different set of rules. If the
+crate is located in the appropriate top-level directory (`tests`, `examples`,
+and so on), no `--module` list will be collected by default. However,
+subdirectories of these directories will be treated as individual binary
+crates: a `main.rs` file will be treated as the root module, and all other
+appropriately named files will be passed as `--module`, using the same
+rules described above.
+
+So if you have an examples directory like this:
+
+```
+examples/
+    foo.rs
+    bar/
+       main.rs
+       baz.rs
+```
+
+This contains two examples, a `foo` example and a `bar` example, and the `bar`
+crate will have `baz.rs` as a submodule.
+
+The reason for this is that today, cargo will treat every file in `tests`,
+`examples`, and `benches` as independent crates, which is a well-motivated
+design. Usually, these are small enough that a single file makes sense.
+However, today, cargo does not make it particularly easy to have tests,
+examples, or benchmarks that are multiple files. This design will create a
+pattern to enable users to do this.
+
+#### Providing your own module list
+
+The target section of the `Cargo.toml` will  gain a new item called `modules`.
+This item is expected to be an array of strings, which are relative paths from
+the cargo manifest directory to files containing Rust source code. If this item
+is specified, cargo will canonicalize these paths and pass them to rustc as the
+`--module` argument when building this target.
+
+## Drawbacks
+
+The RFC authors believe that making mod statements unnecessary is a *net* win,
+but we must acknowledge that it is not a *pure* win. There are several
+advantages that mod statements bring which will not be fully replicated in the
+new system.
+
+Some workflows have been convenienced by the fact that statements need to be
+added to the source code to add new modules to files. For example, it makes it
+easier for users to leave their src directories a little bit dirty while
+working, such as through an incomplete `git stash`. If users wish to comment
+out a module, it can be easier to comment out the `mod` statement than to
+comment out the module file. In general, it enables users to leave code which
+would not compile in their src directory without explicitly commenting it out.
+
+Some users have expressed strong concerns that by deriving the module structure
+from the file system, without making additional syntactic statements, they will
+not be able to as easily find the information they need to navigate and
+comprehend the codebases they are reading or working on. To partly ease their
+concern, the RFC allows users to explicitly specify their module lists at the
+build layer, instead of the source layer. This has some disadvantages, in that
+users may prefer to not have to open the build configuration either.
+
+This will require migrating users away from `mod` statements toward the new
```

Comment:
@arielb1 If rustc could take a file, listing all the modules, instead of fetting all the modules passed via cmd-line arguments, would that appease your concerns? 

This is, for example, how C++ modules work. There one has a `modules.modulemap` file, that maps files, to modules, and the compiler just needs to know where this file is to include all the modules required (compilers can also include specific modules directly). 

Maybe what cargo could do is automatically generate these "module map" files for you, and then call rustc with the path to the temporary module map file. That way, you would be able to copy-paste the command of `cargo build --verbose`. Those wanting to write their own modulemap files, could then do so, and just add a relative path to their `Cargo.toml` to state where this file is. That way, the `Cargo.toml` file remains "cleaner".